### PR TITLE
Filter implicit parameter instances from data-dependencies

### PR DESCRIPTION
**Pull request by cocreature**
Occasionally GHC likes to float implicit parameters to the top-level
making them look like any other typeclass instances. Implicit
parameters are by design not unique so if we try to reconstruct them
we get an error about a duplicate instance.

I haven’t figured out when exactly GHC floats things to the top-level
but I did verify that this testcase breaks without the fix.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/damlc/tests/src/DA/Test/Packaging.hs [hunk 1]
```diff
@@ -1717,6 +1717,49 @@ dataDependencyTests Tools{damlc,repl,validate,davlDar,oldProjDar} = testGroup "D
...
+            , "f = scenario do"
+            , "  p <- getParty \"p\""
+            , "  submit p $ pure ()"
+            , "  submit p $ pure ()"
+            -- This will produce two implicit instances.
+            -- GHC occasionally seems to inline those instances and I don’t understand
+            -- how to reliably stop it from doing this so be careful when changing this.
+            ]
+        withCurrentDirectory (tmpDir </> "dep") $
+            callProcessSilent damlc ["build", "-o", "dep.dar"]
```
Comment (sofiafaro-da) [inline compiler/damlc/tests/src/DA/Test/Packaging.hs:1743]: I wonder if we can make a test that reliably generates an implicit parameter instance?

Or perhaps, add some a check to this test that an implicit parameter instance is in the DAR.

**Review by sofiafaro-da — APPROVED** [review]
The fix looks good!

File: compiler/damlc/tests/src/DA/Test/Packaging.hs (hunk 1 above)
Comment (sofiafaro-da) [reply compiler/damlc/tests/src/DA/Test/Packaging.hs:1743]: Ah, I wonder if we can slip it in to the "Simple DALF" test or something. (It's not so simple...)

Comment (cocreature) [reply compiler/damlc/tests/src/DA/Test/Packaging.hs:1744]: I’ve added an assertion that we have 2 implicit instances. Simple DALF would work as well I guess but this has the advantage that it tests what our users actually hit which seems useful. I still don’t understand how to generate it reliably. Even in this example you can remove the type signature, the inferred type stays the same but now it’s inlined :shrug:

Comment (sofiafaro-da) [reply compiler/damlc/tests/src/DA/Test/Packaging.hs:1744]: Awesome, thank you!

**Review by hurryabit — COMMENTED** [review]
This scares me a bit. But thanks a lot for the quick fix.

**Pull request merged**